jody/hash.go

```diff
@@ -109,6 +114,11 @@ func AddString64(h uint64, s string) uint64 {
 	return h
 }
 
+// AddBytes64 adds the hash of b to the precomputed hash value h.
+func AddBytes64(h uint64, b []byte) uint64 {
+	return AddString64(h, *(*string)(unsafe.Pointer(&b)))
```

Comment:
This should be unnecessary, the compiler _sees_ through the `AddString64` call that the string is not retained by the function and already optimizes the`[]byte -> string` conversion to share the same underlying buffer.

Reply:
Fancy.  You wouldn't happen to have a link to a description of this behavior, would you?  Would love to understand it better.

Reply:
I don't think it's documented because it's not part of the language specs, it's just tribal knowledge. You can verify it with a benchmark and `-benchmem` to count mallocs tho.

Reply:
Heh heh...nice.  Just to make sure I understand, then...  Say I want to hash a byte array.  Then in my client call, I can call `fnv1.HashString64(string(b))`, and the compiler will make sure not to copy `b`?

Reply:
Yep

Reply:
Crazy.

Reply:
Could this be reconsidered? I was about to open a similar PR when I've seen this. The allocation-free []byte to string conversion happens only in some circumstances - namely when the byte array's length is <=32, and even then, the string is on the stack and the data has to be copied.

See https://gist.github.com/mhr3/3cb095e02b759dbb9b605b5ada152bc1

Reply:
Sure, we can reconsider if there's a use case.

The benchmark you shared is puzzling, it seems like the decision is made at runtime in some cases, which is news to me.

This behavior is highly dependent on the compiler version, do you mind sharing which version you ran the benchmarks with?

Reply:
The gist was ran with 1.13.8, but this optimization has been in Go for a very long time (but it is just an optimization, generally speaking doing `string(byteSlice)` does require an alloc).

Reply:
I simplified your example a little to investigate:
```go
package hash_test

import (
	"strings"
	"testing"

	"github.com/segmentio/fasthash/fnv1a"
)

func BenchmarkCopies(b *testing.B) {
	testCases := []string{
		strings.Repeat("a", 8),
		strings.Repeat("a", 64),
		strings.Repeat("a", 128),
	}

	for _, s := range testCases {
		byteArr := []byte(s)

		b.Run("[]byte", func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				fnv1a.HashString64(string(byteArr))
			}
		})
	}
}

```
Still seeing the same behavior in the benchmark:
```
goos: darwin
goarch: amd64
pkg: hash/test
BenchmarkCopies
BenchmarkCopies/[]byte
BenchmarkCopies/[]byte         	126141844	         9.52 ns/op	       0 B/op	       0 allocs/op
BenchmarkCopies/[]byte#01
BenchmarkCopies/[]byte#01      	15601798	        73.1 ns/op	      64 B/op	       1 allocs/op
BenchmarkCopies/[]byte#02
BenchmarkCopies/[]byte#02      	 8390252	       140 ns/op	     128 B/op	       1 allocs/op
PASS
ok  	hash/test	5.197s
```
The escape analysis indicates that the conversion of `string(byteArr)` shouldn't cause a heap allocation:
```
./hash.go:22:31: BenchmarkCopies.func1 capturing by value: byteArr (addr=false assign=false width=24)
./hash.go:10:6: cannot inline BenchmarkCopies: unhandled op RANGE
./hash.go:20:19: cannot inline BenchmarkCopies.func1: unhandled op FOR
./hash.go:22:23: inlining call to fnv1a.HashString64 func(string) uint64 { return fnv1a.AddString64(uint64(14695981039346656037), fnv1a.s) }
./hash.go:20:19: func literal escapes to heap:
./hash.go:20:19:   flow: {heap} = &{storage for func literal}:
./hash.go:20:19:     from func literal (spill) at ./hash.go:20:19
./hash.go:20:19:     from b.Run("[]byte", func literal) (call parameter) at ./hash.go:20:8
./hash.go:10:22: parameter b leaks to {heap} with derefs=0:
./hash.go:10:22:   flow: {heap} = b:
./hash.go:10:22:     from b.Run("[]byte", func literal) (call parameter) at ./hash.go:20:8
./hash.go:18:20: ([]byte)(s) escapes to heap:
./hash.go:18:20:   flow: byteArr = &{storage for ([]byte)(s)}:
./hash.go:18:20:     from ([]byte)(s) (spill) at ./hash.go:18:20
./hash.go:18:20:     from byteArr := ([]byte)(s) (assign) at ./hash.go:18:11
./hash.go:18:20:   flow: {storage for func literal} = byteArr:
./hash.go:18:20:     from func literal (captured by a closure) at ./hash.go:20:19
./hash.go:10:22: leaking param: b
./hash.go:20:24: b does not escape
./hash.go:11:23: []string literal does not escape
./hash.go:18:20: ([]byte)(s) escapes to heap
./hash.go:20:19: func literal escapes to heap
./hash.go:22:30: string(byteArr) does not escape
```
I'm confused why this results in a heap allocation 🤔

Reply:
Whether the variable escapes doesn't matter, `string -> []byte`, as well as `[]byte -> string` conversions need to copy the data, otherwise Go wouldn't be able to keep its promise of immutable strings, anytime the conversion doesn't require a copy is just an edge-case optimization.

Reply:
My understanding was that the copy was eluded when the compiler could determine that the program was not mutating the value, either I was wrong or this was changed at some point in a release 🤷‍♂️ 

Anyways, happy to re-open 👍

Reply:
Thanks, so is anything missing here?

Reply:
Seems like CI isn't passing, I would also like to get @stevevls to give it a look since he was the original author.

Reply:
I'd be happy to bring this back.  In addition to fixing CI,  I think we'd want to incorporate the loop unrolling from #9.